### PR TITLE
Add InRange converter and re.fullmatch converter

### PR DESCRIPTION
**Pull request by TurnrDev**

### CODE REVIEW

File: discord/ext/converters/regex.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+class MatchConverter(commands.Converter):
+    def __init__(self, *, pattern: str, exception: str = "No match found.") -> None:
+        self.pattern = pattern
+        self.exception = exception
+
+    async def convert(self, ctx, argument: str) -> str:
+        match: Union[re.Match, None] = re.match(self.pattern, argument)
+        if match is None:
+            raise commands.BadArgument(self.exception)
+        return argument
```
Comment (TurnrDev) [inline discord/ext/converters/regex.py]: I should probably return the match object instead

File: discord/ext/converters/regex.py [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+from discord.ext import commands
+
+
+class MatchConverter(commands.Converter):
+    def __init__(self, *, pattern: str, exception: str = "No match found.") -> None:
+        self.pattern = pattern
+        self.exception = exception
+
+    async def convert(self, ctx, argument: str) -> str:
+        match: Union[re.Match, None] = re.match(self.pattern, argument)
```
Comment (Gobot1234) [inline discord/ext/converters/regex.py]: Using `Union[..., None]` is not generally the way you should do that normally you should use `Optional[...]`

File: discord/ext/converters/range.py [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+
+from discord.ext import commands
+
+
+class InRangeConverter(commands.Converter):
+    def __init__(self, *, range_tuple: Tuple) -> None:
+        self.range = range(*range_tuple)
+
+    async def convert(self, ctx, argument: str) -> str:
+        if argument in self.range:
```
Comment (ShineyDev) [inline discord/ext/converters/range.py:14]: A `str` will never exist in a `range`.

```suggestion
        if argument in string.digits and int(argument) in self.range:
```

File: discord/ext/converters/range.py [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+            return argument
+        raise commands.BadArgument("{argument} not in range".format(argument=argument))
```
Comment (ShineyDev) [inline discord/ext/converters/range.py]:
```suggestion
        raise commands.BadArgument("{0} not in range".format(argument))
```

File: discord/ext/converters/regex.py [hunk 5]
```diff
@@ -0,0 +1,16 @@
+import re
+from typing import Union
+
+from discord.ext import commands
+
+
+class MatchConverter(commands.Converter):
+    def __init__(self, *, pattern: str, exception: str = "No match found.") -> None:
```
Comment (ShineyDev) [inline discord/ext/converters/regex.py]: You have the right idea but this is silly.

```suggestion
    def __init__(self, pattern: str) -> None:
```

It isn't a terrible idea to allow passing RegEx flags here too.

File: discord/ext/converters/regex.py (hunk 2 above)
Comment (ShineyDev) [inline discord/ext/converters/regex.py]: I recall Python's `re.match` is borderline useless, probably use `re.fullmatch` instead.

File: discord/ext/converters/regex.py (hunk 1 above)
Comment (ShineyDev) [inline discord/ext/converters/regex.py]: [You're right](https://github.com/Ext-Creators/discord-ext-converters/pull/1#discussion_r499825306). This should return a `re.Match`.

**Review by ShineyDev — CHANGES_REQUESTED** [review]

File: discord/ext/converters/range.py [hunk 6]
```diff
@@ -0,0 +1,13 @@
+from typing import Tuple
+
+from discord.ext import commands
+
+
+class InRangeConverter(commands.Converter):
+    def __init__(self, *, range_tuple: Tuple) -> None:
```
Comment (Gobot1234) [inline discord/ext/converters/range.py]: In my opinion this shouldn't take a tuple with the values, but instead just directly pass the arguments to `self.range`.

Consider - `InRangeConverter(range_tuple=(1, 2, 3))` 
This seems rather extraneous as it is clear from context what the arguments mean.
I propose - `InRangeConverter(1, 2, 3)` as the better syntax for this.

Comment (ShineyDev) [reply discord/ext/converters/range.py]: SGTM. Either way, this should *not* be a keyword argument - good catch.

Comment (TurnrDev) [reply discord/ext/converters/range.py]: Good point!

File: discord/ext/converters/range.py (hunk 3 above)
Comment (TurnrDev) [reply discord/ext/converters/range.py:14]: :man_facepalming: I don't know how I made such a rookie mistake. :fearful:

File: discord/ext/converters/regex.py (hunk 5 above)
Comment (TurnrDev) [reply discord/ext/converters/regex.py]: ~~Surely passing all args and kwargs to match should be best here~~

**Comment by TurnrDev** [comment]
Thank you for your feedback. I've fixed the issues and a few others I spotted as I went. :)

File: discord/ext/converters/regex.py [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+        match: Optional[re.Match] = re.fullmatch(
+            self.pattern, argument, flags=self.flags
+        )
+        if match:
+            raise commands.BadArgument(
+                "{0} does not match in the regex pattern ({1})".format(
+                    argument, self.pattern
+                )
+            )
+        return match
```
Comment (Gobot1234) [inline discord/ext/converters/regex.py]: This doesn't seem right unless I'm missing something

Comment (TurnrDev) [reply discord/ext/converters/regex.py]: You're right, I meant if not match :man_facepalming:

File: discord/ext/converters/range.py [hunk 8]
```diff
@@ -0,0 +1,13 @@
+from typing import Tuple
+
+from discord.ext import commands
+
+
+class InRangeConverter(commands.Converter):
+    def __init__(self, *args, **kwargs) -> None:
+        self.range = range(*args, **kwargs)
```
Comment (ShineyDev) [inline discord/ext/converters/range.py]: `range` doesn't take any keyword arguments.

```suggestion
        self.range = range(*args)
```

File: discord/ext/converters/range.py [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+
+    async def convert(self, ctx, argument: str) -> int:
+        if argument.isnumeric() and int(argument) in self.range:
```
Comment (ShineyDev) [inline discord/ext/converters/range.py]: `str.isnumeric` does not mean `int(str)` will not fail. This should follow zen and ask forgiveness instead.

File: discord/ext/converters/regex.py [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+
+
+class MatchConverter(commands.Converter):
+    def __init__(self, pattern, flags) -> None:
+        self.pattern = pattern
+        self.flags = flags
+
+    async def convert(self, ctx, argument: str) -> re.Match:
+        match: Optional[re.Match] = re.fullmatch(
+            self.pattern, argument, flags=self.flags
```
Comment (ShineyDev) [inline discord/ext/converters/regex.py]: `flags` is not a keyword argument.

```suggestion
            self.pattern, argument, self.flags
```

File: discord/ext/converters/regex.py [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+        )
+        if match is None:
+            raise commands.BadArgument(
+                "{0} does not match in the regex pattern ({1})".format(
```
Comment (ShineyDev) [inline discord/ext/converters/regex.py]: Adding the pattern string to the exception only makes it more convoluted.

**Review by ShineyDev — CHANGES_REQUESTED** [review]
Getting there.

File: discord/ext/converters/regex.py (hunk 10 above)
Comment (TurnrDev) [reply discord/ext/converters/regex.py]: [It is](https://docs.python.org/3/library/re.html#re.fullmatch)?

File: discord/ext/converters/regex.py (hunk 11 above)
Comment (TurnrDev) [reply discord/ext/converters/regex.py]: What would you recommend as wording?

**Review by ShineyDev — APPROVED** [review]
LGTM. I'll request a closing review from @NCPlayz.

**Review by Gobot1234 — APPROVED** [review]

**Review by NCPlayz — APPROVED** [review]
Thanks for the PR. I'll make any required changes to work with #2 afterwards.

**Pull request merged**